File: June-LeetCoding-Challenge-2020/Week1/1029-two-city-scheduling.js
/**
 * @param {number[][]} costs
 * @return {number}
 * Runtime: 68 ms, faster than 39.39% of JavaScript online submissions for Two City Scheduling.
 * Memory Usage: 33.7 MB, less than 100.00% of JavaScript online submissions for Two City Scheduling.
 * TC: O(NLogN), where N is the number of elements in costs
 * SC: O(N), where N is the number of elements in costs
 */
var twoCitySchedCost = function(costs) {
    const diff = costs.map(cost => cost[1] - cost[0]);

    diff.sort((cost0, cost1) => cost0 - cost1);

    let possibleSmallestDiff = 0;
    for (let index = 0; index < diff.length / 2; index++) {
        possibleSmallestDiff += diff[index];
    }


    return costs.reduce((acc, cur) => acc + cur[0], possibleSmallestDiff);
};

File: June-LeetCoding-Challenge-2020/Week1/1029-two-city-scheduling.py
class Solution:
    def twoCitySchedCost(self, costs: List[List[int]]) -> int:
        costs.sort(key=lambda x: x[0]-x[1])
        # print(costs)
        total = 0
        n = len(costs)//2 # we send ppl/2 exactly to each city
        
        for i in range(n):
            total += costs[i][0] + costs[i+n][1]
            
        return total
    # TC: O(nlogn) + O(n) = O(nlogn), where sort() takes O(nlogn)
    # SC: O(1) 
    # Runtime: 44 ms, faster than 49.91%
    # Memory: 14.1 MB, less than 7.69%